### PR TITLE
[SPARK-26218][SQL][Follow up] Fix the corner case when casting float to Integer.

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?
When spark.sql.ansi.enabled is true, for the statement:
```
select cast(cast(2147483648 as Float) as Integer) //result is 2147483647
```
Its result is 2147483647 and does not throw `ArithmeticException`.

The root cause is that, the below code does not work for some corner cases.
https://github.com/apache/spark/blob/94fc0e3235162afc6038019eed6ec546e3d1983e/sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala#L129-L141

For example:

image

In this PR, I fix it by comparing Math.floor(x) with Int.MaxValue directly.

### Why are the changes needed?
Result corrupt.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?

Added Unit test.

### CODE REVIEW

**Comment by turboFei** [comment]
For long:
image

**Comment by turboFei** [comment]
For pgsql:
image
For teradata:
image

**Comment by turboFei** [comment]
cc @wangyum  @cloud-fan @gengliangwang

**Comment by cloud-fan** [comment]
OK to test

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -3384,6 +3384,17 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
       assert(exp.getMessage.contains("Resources not found"))
     }
   }
+
+  test("SPARK-26218: Fix the corner case when casting float to Integer") {
+    withSQLConf(SQLConf.ANSI_ENABLED.key -> "true") {
+      intercept[ArithmeticException](
+        sql("SELECT CAST(CAST(2147483648 as FLOAT) as Integer)").explain()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: let's call `collect` to make the execution code path simple

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -3384,6 +3384,17 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
...
+      )
+      intercept[ArithmeticException](
+        sql("SELECT CAST(CAST(2147483648 as DOUBLE) as Integer)").show()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: ditto

**Review by cloud-fan — APPROVED** [review]

**Comment by srowen** [comment]
Hm, but:
```
scala> (Int.MaxValue.toFloat+1).toInt
res13: Int = 2147483647

scala> (Int.MaxValue.toFloat+1).toInt == Int.MaxValue
res14: Boolean = true
```
Those values do correctly cast to an int. The cast does lose precision of course, but according to Scala/Java, the result is correct, no?

**Comment by turboFei** [comment]
> Hm, but:
> 
> ```
> scala> (Int.MaxValue.toFloat+1).toInt
> res13: Int = 2147483647
> 
> scala> (Int.MaxValue.toFloat+1).toInt == Int.MaxValue
> res14: Boolean = true
> ```
> 
> Those values do correctly cast to an int. The cast does lose precision of course, but according to Scala/Java, the result is correct, no?

Yes, the behavior is consistent with Scala/Java, it seems that if the value exceeds Int.Max,  cast it to Int is Int.Max.
But when spark.sql.ansi.enabled is true, we should throw exception to keep consistent with ansi.

**Comment by srowen** [comment]
Is this code path only used for ANSI mode? and is that defined by ANSI? I wouldn't expect the result of the cast to retain that much accuracy. You're not in general going to get the same int out when the int is large, after the round-trip - right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala [hunk 3]
```diff
@@ -121,8 +121,8 @@ object FloatExactNumeric extends FloatIsFractional {
   private def overflowException(x: Float, dataType: String) =
     throw new ArithmeticException(s"Casting $x to $dataType causes overflow")
 
-  private val intUpperBound = Int.MaxValue.toFloat
-  private val intLowerBound = Int.MinValue.toFloat
+  private val intUpperBound = Int.MaxValue
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: Hm, I'm also not clear how this helps - won't it just promote to a float in the comparison below anyway?
Do we want floorDiv, etc, instead?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: `Math.floor` returns double, so it's promoted to double

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala [hunk 4]
```diff
@@ -121,8 +121,8 @@ object FloatExactNumeric extends FloatIsFractional {
...
+  private val intLowerBound = Int.MinValue
   private val longUpperBound = Long.MaxValue.toFloat
   private val longLowerBound = Long.MinValue.toFloat
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: seems we can remove `toFloat` here too? also the `toDouble` in `DoubleExactNumeric`. They will be promoted anyway.

Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: Agree. It looks more gracefully with consistent style.

**Comment by turboFei** [comment]
> Is this code path only used for ANSI mode? and is that defined by ANSI? I wouldn't expect the result of the cast to retain that much accuracy. You're not in general going to get the same int out when the int is large, after the round-trip - right?

Yes, it would be only invoked for ANSI mode.
Ansi would throw exception when overflow.
I have attached the relative behavior of pgsql and teradata above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala (hunk 3 above)
Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: As mentioned by cloud-fan, it seems that an cast int to float, then casted to double is not same with casting it to double directly.
image

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: it's true
```
scala> Int.MaxValue.toDouble
res2: Double = 2.147483647E9

scala> Int.MaxValue.toFloat.toDouble
res3: Double = 2.147483648E9
```

**Comment by cloud-fan** [comment]
retest this please

**Comment by turboFei** [comment]
Thanks for your review. May we need close this PR.
Float is not an accurate type and its scale is only 8 bits.
And the Math.floor(floatValue)  operation would also cast this float value to a double value, so it's reasonable to compare with Math.floor(floatValue) with Int.MaxValue.toFloat.

Thanks for your review again.

**Comment by cloud-fan** [comment]
We are talking about SQL semantic not IEEE floating number definition.

For pgsql
```
cloud0fan=# SELECT CAST(CAST(2147483648 as FLOAT) as Int);
ERROR:  integer out of range
```

I think the fix makes sense.

**Comment by srowen** [comment]
(OK I'm into the idea, yes)

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
@turboFei can you fix the conflicts?

**Comment by kiszk** [comment]
ping @turboFei

**Comment by turboFei** [comment]
Thanks for your reminder. 
As mentioned above, 
> Thanks for your review. May we need close this PR.
> Float is not an accurate type and its scale is only 8 bits, which means that 21234567890f == 21234567800f.
> And the Math.floor(floatValue) operation would also cast this float value to a double value, so it's reasonable to compare with Math.floor(floatValue) with Int.MaxValue.toFloat.
> 
> Thanks for your review again.

I think I need close this PR. 
I will also try to find other solution.

**Comment by cloud-fan** [comment]
the PR looks good as it is. Do you hit some test failures that you are not able to fix?

**Comment by turboFei** [comment]
Yes, Here is a test failure.
image

```
scala> 2147483647f.toDouble < 2147483647.toDouble
res2: Boolean = false
```

**Comment by cloud-fan** [comment]
If you look at the pgsql result, the new result is actually corrected: https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/expected/float4.out#L299

Can you just re-generate the answer files? You can look at the doc of `SQLQueryTestSuite` to see how to do it.

**Comment by turboFei** [comment]
I will do it later, thanks.

**Comment by cloud-fan** [comment]
unfortunately it conflicts. Can you fix it by git rebase? thanks!

**Comment by turboFei** [comment]
Except Int.MaxValue, the max Integer i, which satisfies i.toFloat.toInt == i, is 2147483520.

And it has been added into the UT(query-34/float4.sql). So, I just remove query-35.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/float4.sql [hunk 5]
```diff
@@ -106,7 +106,6 @@ SELECT smallint(float('32767.6'));
 SELECT smallint(float('-32768.4'));
 SELECT smallint(float('-32768.6'));
 SELECT int(float('2147483520'));
-SELECT int(float('2147483647'));
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/float4.sql]: These tests are copied from pgsql and we shouldn't change it. We just need to re-generate the answer file and keep the actual result as it is.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/float4.sql [hunk 6]
```diff
@@ -106,7 +106,7 @@ SELECT smallint(float('32767.6'));
...
+SELECT int(float('2147483392'));
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/float4.sql]: let's NOT change the pgsql tests. They are used to verify the difference between Spark and pgsql. We should respect the test result, whatever it is.

**Comment by cloud-fan** [comment]
The last 2 commits are empty and just to trigger jenkins. The last effective commit has passed tests. I'm merging it to master/3.0, thanks!
